cl/citations/find_citations.py

```diff
@@ -377,66 +229,156 @@ def add_defendant(citation, words):
-        page = int(page)
+    # Get page
+    page = parse_page(page)
+    if not page:
+        return None
+
+    # Return FullCitation
+    return FullCitation(
+        reporter,
+        page,
+        volume,
+        reporter_found=reporter,
+        reporter_index=reporter_index,
+    )
+
+
+def extract_shortform_citation(words, reporter_index):
+    """Given a list of words and the index of a federal reporter, look before
+    and after to see if this is a short form citation. If found, construct
+    and return a ShortformCitation object.
+
+    Shortform 1: Adarand, 515 U.S., at 241
+    Shortform 2: 515 U.S., at 241
+    """
+    # Don't check if we are the beginning of a string
+    if reporter_index == 0:
+        return None
+
+    # Get volume
+    volume = strip_punct(words[reporter_index - 1])
+    if volume.isdigit():
+        volume = int(volume)
     else:
-        if isroman(page):
-            # Some places like Nebraska have Roman numerals, e.g. in
-            # '250 Neb. xxiv (1996)'. No processing needed.
-            pass
-        elif re.match(r"\d{1,6}[-]?[a-zA-Z]{1,6}", page):
-            # Some places, like Connecticut, have pages like "13301-M".
-            # Other places, like Illinois have "pages" like "110311-B".
-            pass
-        else:
-            # Not Roman, and not a weird connecticut page number. Thus a bad
-            # value. Abort.
+        # No volume, therefore not a valid citation
+        return None
+
+    # Get page
+    page = parse_page(words[reporter_index + 2])
+    if not page:
+        # There might be a comma in the way, so try one more index
+        page = parse_page(words[reporter_index + 3])
+        if not page:
+            # No page, therefore not a valid citation
             return None
 
-    return Citation(
+    # Get antecedent
+    antecedent_guess = words[reporter_index - 2]
+    if antecedent_guess == ",":
+        antecedent_guess = words[reporter_index - 3] + ","
+
+    # Get reporter
+    reporter = words[reporter_index]
+
+    # Return ShortformCitation
+    return ShortformCitation(
         reporter,
         page,
         volume,
+        antecedent_guess,
         reporter_found=reporter,
         reporter_index=reporter_index,
     )
 
 
+def extract_supra_citation(words, supra_index):
+    """Given a list of words and the index of a supra token, look before
+    and after to see if this is a supra citation. If found, construct
+    and return a SupraCitation object.
+
+    Supra 1: Adarand, supra, at 240
+    Supra 2: Adarand, 515 supra, at 240
+    Supra 3: Adarand, supra, somethingelse
+    Supra 4: Adrand, supra. somethingelse
+    """
+    # Get volume
+    volume = None
+
+    # Get page
+    page = parse_page(words[supra_index + 2])
+
+    # Get antecedent
+    antecedent_guess = words[supra_index - 1]
+    if antecedent_guess.isdigit():
+        volume = int(antecedent_guess)
+        antecedent_guess = words[supra_index - 2]
```

Comment:
Am I wrong or will these negative indexes either crash or just plain do weird things if supra_index = 0 or 1?

Reply:
Yes, I think you're right, but my assumption was that `supra_index` would *never* be 0 or 1, since it would be impossible for a supra citation to appear at the beginning of a document. (Since by definition, "supra" would have to refer to something *prior*.) However, I know that sane input is never guaranteed, and it's easy to add a check for this, so I will fix these 👍

Reply:
These are on my radar because I know our regular parser crashes for this reason. But it is more likely to crash because lots of reporters can be the first word, but yeah, protecting against this will probably save us trouble.

Reply:
Fixed in d024203